### PR TITLE
match: Pass streamInfo as input to Http matching data

### PR DESCRIPTION
**Pull request by tyxia**
Commit Message:
With streamInfo as the input to Http matching data, it not only fulfills the existing functionality provided by ConnectionInfoProvider (ConnectionInfoProvider is got from stream_info.downstreamAddressProvider()), it but also provides more info that is available in streamInfo, via, for example, dynamic_metatdata and filter state.
The use case could be : the matching library like CEL expression could match on, for example, client_ip, upstream cluster name, etc that can be retrieved  from stream info.

Signed-off-by: tyxia [tyxia@google.com](mailto:tyxia@google.com)

### CODE REVIEW

File: test/common/http/matching/inputs_test.cc [hunk 1]
```diff
@@ -4,19 +4,26 @@
...
-TEST(MatchingData, HttpRequestHeadersDataInput) {
-  HttpRequestHeadersDataInput input("header");
-  Network::ConnectionInfoSetterImpl connection_info_provider(
+// Helper function to create the stream info.
+StreamInfo::StreamInfoImpl createStreamInfo() {
+  auto connection_info_provider = std::make_shared<Network::ConnectionInfoSetterImpl>(
       std::make_shared<Network::Address::Ipv4Instance>(80),
       std::make_shared<Network::Address::Ipv4Instance>(80));
-  HttpMatchingDataImpl data(connection_info_provider);
+  return {Http::Protocol::Http2, Event::GlobalTimeSystem().timeSystem(), connection_info_provider};
```
Comment (kyessenov) [inline test/common/http/matching/inputs_test.cc:21]: Can you use a constructor? This is error-prone when fields change in a struct.

**Review by kyessenov — COMMENTED** [review]
LGTM after a suggestion, leave to @yanavlasov to complete the review.

File: test/common/http/matching/inputs_test.cc (hunk 1 above)
Comment (tyxia) [reply test/common/http/matching/inputs_test.cc]: Thanks for review. Done.

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**